src/lib/header-bidding/prebid/modules/analyticsAdapter.ts

```diff
@@ -91,235 +69,255 @@ function logEvents(events: EventData[]): void {
-	setSafely(event, 'ttr', duration);
-	return [event];
-};
+	},
+	[EVENTS.BID_REQUESTED]: (_, args) => {
+		if (args.bids) {
+			return args.bids.map((bid) => {
+				const event = createEvent({
+					ev: 'request',
+					n: args.bidderCode,
+					sid: bid.adUnitCode,
+					bid: bid.bidId,
+					st: args.start,
+				});
+
+				return event;
+			});
+		}
+		return null;
+	},
+	[EVENTS.BID_RESPONSE]: (_, args) => {
+		if (args.statusMessage === 'Bid available') {
+			const event = createEvent({
+				ev: 'response',
+				n: getBidderCode(args),
+				bid: args.requestId,
+				sid: args.adUnitCode,
+				cpm: args.cpm,
+				pb: args.pbCg,
+				cry: args.currency,
+				net: args.netRevenue,
+				did: args.adId,
+				cid: args.creativeId,
+				sz: args.size,
+				ttr: args.timeToRespond,
+				lid: args.dealId,
+				dsp: args.meta?.networkId,
+				adv: args.meta?.buyerId,
+				bri: args.meta?.brandId,
+				brn: args.meta?.brandName,
+				add: args.meta?.clickUrl,
+			});
+
+			return [event];
+		}
+		return null;
+	},
+	[EVENTS.NO_BID]: (adapter, args) => {
+		const duration = Date.now() - (adapter.context?.auctionTimeStart ?? 0);
+		const event = createEvent({
+			ev: 'nobid',
+			n: args.bidder ?? args.bidderCode,
+			bid: args.bidId ?? args.requestId,
+			sid: args.adUnitCode,
+			aid: args.auctionId,
+			ttr: duration,
+		});
 
-class AnalyticsQueue {
-	private queue: EventData[] = [];
+		return [event];
+	},
+	[EVENTS.AUCTION_END]: (adapter, args) => {
+		const duration = Date.now() - (adapter.context?.auctionTimeStart ?? 0);
+		const event = createEvent({
+			ev: 'end',
+			aid: args.auctionId,
+			ttr: duration,
+		});
 
-	push(event: EventData | EventData[]): void {
-		if (Array.isArray(event)) {
-			this.queue.push(...event);
-		} else {
-			this.queue.push(event);
-		}
-	}
+		return [event];
+	},
+	[EVENTS.BID_WON]: (_, args: BidArgs) => {
+		const event = createEvent({
+			ev: 'bidwon',
+			aid: args.auctionId,
+			bid: args.requestId,
+		});
+		return [event];
+	},
+} as const satisfies Record<string, Handler>;
 
-	popAll(): EventData[] {
-		const result = this.queue;
-		this.queue = [];
-		return result;
-	}
+type AnalyticsPayload = {
+	v: number;
+	pv: string;
+	hb_ev: EventData[];
+};
 
-	/**
-	 * For test/debug purposes only
-	 */
-	peekAll(): EventData[] {
-		return this.queue;
-	}
+// Check if the payload is valid
+function isPayloadValid(events: EventData[]): boolean {
+	return (
+		(events[0] && (events[0].ev === 'init' || events[0].ev === 'bidwon')) ??
+		false
+	);
+}
 
-	init(): void {
-		this.queue = [];
+const createPayload = (events: EventData[], pv: string): AnalyticsPayload => {
+	const payload = {
+		v: VERSION,
+		pv,
+		hb_ev: events,
+	};
+	if (!isPayloadValid(events)) {
+		reportError(
+			new Error('Invalid analytics payload'),
+			'commercial',
+			{},
+			{
+				payload,
+			},
+		);
 	}
-}
+	return payload;
+};
 
 const analyticsAdapter = Object.assign(adapter({ analyticsType: 'endpoint' }), {
+	sendPayload: async (
```

Comment:
the only reason this function is in the adapter is so it can be mocked for testing.